### PR TITLE
Add str::Lines::as_str

### PR DESCRIPTION
**Pull request by jplatte**
~~Blocked on #95644~~, for now I'm using the same API as available on `SplitTerminator` but this should probably not be merged as-is since that API is changing.

Just wanted to open this now as a means to propose having the same "remainder of non-iterated parts as a contiguous `&str`" API on `Lines` as on `Split*` (tracked in #77998).

I'm also curious whether you think it makes sense to implement it like this (reaching into the inner `SplitTerminator` of `Map`) or whether the `Map` should maybe be replaced? I found the code a little hard to follow at first given it uses `Map` instead of just (?) doing the same thing manually in the `Iterator` implementation.

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
fyi: #95644 got merged, so this is not blocked on it anymore (however this needs a rebase — #95644 changed the APIs that this PR uses).

**Comment by Dylan-DPC** [comment]
@jplatte any updates on this pr?

**Comment by jplatte** [comment]
Don't have much time right now, since it's a small thing I might be able to get to it soon though. If somebody else wants to take over that would be fine by me too, just please ping me so the work isn't duplicated.

**Comment by jplatte** [comment]
Closing in favor of #107464. Thanks WaffleLapkin!
